hep/proposals/0001-decorators.md

```diff
@@ -0,0 +1,313 @@
+    # inject outputs based on inputs.
+    output: MyOutput = MyOutput(my_input)
+
+    # template is a special variable that allows you to reference the template itself and modify
+    # its attributes. This is especially useful when trying to reference input params in args.
+    # The hio.name and hio.path functions will be able to inspect the annotation of the given variable
+    # to return the correct Argo template syntax substitution.
+    template.args = [f"echo {hio.name(my_input.user)} > {hio.path(output.container_greeting)}"]
+    return output
+```
+
+## Template Refs
+
+```python
+from hera.workflows import WorkflowTemplate, ClusterWorkflowTemplate
+import hera.workflows.io as hio
+
+# We start by defining our Workflow Template
+wt = WorkflowTemplate(name="my-template")
+
+# We will be referring to "another-workflow-template", to demonstrate two WorkflowTemplates can be written in the same package
+awt = WorkflowTemplate(name="another-workflow-template")
+
+# We assume we can generate Python code for this (Cluster) Workflow Template from e.g `hera generate stubs external-workflow-template`.
+# Ultimately, awt and ewt will behave the same way when their templates are invoked in dag/steps functions.
+ewt = ClusterWorkflowTemplate(name="external-workflow-template")
+
+@awt.script  # Adds a new script template to a workflow template called "another-workflow-template"
+def setup() -> hio.Output:
+    return hio.Output(result="Setting things up")
+
+class ConcatInput(hio.Input):
+    word_a: str
+    word_b: str
+
+class ConcatOutput(hio.Output):
+    value: str
+
+# We assume we can autogenerate stubs such as these for templates in "external-workflow-template"
+@ewt.script
+def concat(concat_input: ConcatInput) -> ConcatOutput:
+    ...
+    # Note: we can optionally return hio.output(concat_input)
+    # again here to run "local" mock versions of this template ref
+    # this will be useful in local testing
+
+# In the case of kebab-case template names or other details, we can pass the extra info to the decorator:
+@ewt.script(name="my-concat-function")
+def my_concat_function(concat_input: ConcatInput) -> ConcatOutput:
+    ...
+
+
+class WorkerInput(hio.Input):
+    value_a: str
+    value_b: str
+
+class WorkerOutput(hio.Output):
+    value: str
+
+
+@wt.entrypoint
+@wt.dag
+def worker(worker_input: WorkerInput) -> WorkerOutput:
+    # Here, the call to `setup` will create a template ref task since it belongs
+    # in a different workflow template even though it is defined inside this file
+    setup_task = setup()
+
+    # concat serves as a stub, and will also become a template ref task
+    task_a = concat(ConcatInput(word_a=worker_input.value_a, word_b=setup_task.result))
+    task_b = concat(ConcatInput(word_a=worker_input.value_b, word_b=setup_task.result))
+    final_task = concat(ConcatInput(word_a=task_a.value, word_b=task_b.value))
+    return WorkerOutput(value=final_task.value)
```

Comment:
Since this implicitly sets up the dependencies based on I/O it is now equivalent to a Steps context in terms of the implicit dependencies. That might cause some conceptual tension. Can you explain a bit about how this does not conflict with Steps?

Reply:
It's not equivalent. Steps will still have a parallel context manager and will be executed in sequence. For DAGs, some of the dependencies can be implicitly discovered by us, but the user can also specify them.

The same argument can be put forward for steps vs dags in Argo Workflows. You can convert every DAG into step and the choice is largely upto the user.

Reply:
Added a parallel steps example for more colour, and explained the auto-dag-construction in the DAG example. Agree with Sam that it comes down to Argo Workflows itself and the user's preference